pkg/service/cluster/service.go

```diff
@@ -609,57 +621,102 @@ func (s *Service) GetSession(ctx context.Context, clusterID uuid.UUID) (session
 		}
 	}
 
-	scyllaCluster := gocql.NewCluster(sessionHosts...)
-	cqlPort := ni.CQLPort()
+	cfg := gocql.NewCluster(sessionHosts...)
+	if err := s.extendClusterConfigWithAuthentication(clusterID, ni, cfg); err != nil {
+		return session, err
+	}
+	if err := s.extendClusterConfigWithTLS(ctx, clusterID, ni, cfg); err != nil {
+		return session, err
+	}
+
+	return gocqlx.WrapSession(cfg.CreateSession())
+}
+
+// SingleHostSessionFunc returns single host CQL session for given cluster ID.
+type SingleHostSessionFunc func(ctx context.Context, clusterID uuid.UUID, host string) (gocqlx.Session, error)
+
+// GetSingleHostSession returns single host CQL session to provided cluster.
+// It's useful when we need to ensure that CQL queries are routed to given host.
+func (s *Service) GetSingleHostSession(ctx context.Context, clusterID uuid.UUID, host string) (session gocqlx.Session, err error) {
```

Comment:
There is no need for introducing code duplication.
The signature of `GetSession` can be changed to `func (s *Service) GetSession(ctx context.Context, clusterID uuid.UUID, opts ...SessionOpt) (session gocqlx.Session, err error) {`

Then,
```
type SessionOpt interface {
	apply(*sessionConfig)
}

type sessionConfig struct {
	host        string
	singleHost  bool
}

type singleHostOpt struct {
	host string
}

func (o singleHostOpt) apply(cfg *sessionConfig) {
	cfg.singleHost = true
	cfg.host = o.host
}

func WithSingleHost(host string) SessionOpt {
	return singleHostOpt{host: host}
}
```

and GetSession body:
```
func (s *Service) GetSession(ctx context.Context, clusterID uuid.UUID, opts ...SessionOpt) (session gocqlx.Session, err error) {
	s.logger.Info(ctx, "Get session", "cluster_id", clusterID)

	client, err := s.CreateClientNoCache(ctx, clusterID)
	if err != nil {
		return session, errors.Wrap(err, "get client")
	}
	defer logutil.LogOnError(ctx, s.logger, client.Close, "Couldn't close scylla client")

	config := sessionConfig{}
	for _, opt := range opts {
		opt.apply(&config)
	}

	var ni *scyllaclient.NodeInfo
	if config.singleHost {
		s.logger.Info(ctx, "Get single host session", "host", config.host)
		ni, err = client.NodeInfo(ctx, config.host)
		if err != nil {
			return session, errors.Wrap(err, "fetch node info")
		}
	} else {
		ni, err = client.AnyNodeInfo(ctx)
		if err != nil {
			return session, errors.Wrap(err, "fetch node info")
		}
	}

	var sessionHosts []string
	if config.singleHost {
		sessionHosts = []string{ni.CQLAddr(config.host)}
	} else {
		sessionHosts, err = GetRPCAddresses(ctx, client, client.Config().Hosts)
		if err != nil {
			s.logger.Info(ctx, "GetSession", "err", err)
			if errors.Is(err, ErrNoRPCAddressesFound) {
				return session, err
			}
		}
	}

	cfg := gocql.NewCluster(sessionHosts...)
	if config.singleHost {
		cfg.HostFilter = gocql.WhiteListHostFilter(sessionHosts...)
		cfg.DisableInitialHostLookup = true
	}

	if err := s.extendClusterConfigWithAuthentication(clusterID, ni, cfg); err != nil {
		return session, err
	}
	if err := s.extendClusterConfigWithTLS(ctx, clusterID, ni, cfg); err != nil {
		return session, err
	}

	return gocqlx.WrapSession(cfg.CreateSession())
}
```

Reply:
That's a good idea, but I'm not convinced to the proposed format.
It adds a lot of similar types and requires a lot of (3!) if statements in the method body, which makes it less comprehensible. I would be more in favor of something like:
```go
type SessionConfigOption func(ctx context.Context, clusterID uuid.UUID, client *scyllaclient.Client, cfg *gocql.ClusterConfig) error

func SingleHostSessionOpt(host string) SessionConfigOption {
	return func(ctx context.Context, clusterID uuid.UUID, client *scyllaclient.Client, cfg *gocql.ClusterConfig) error {
		ni, err := client.NodeInfo(ctx, host)
		if err != nil {
			return errors.Wrapf(err, "fetch node (%s) info", host)
		}
		cqlAddr := ni.CQLAddr(host)
		cfg.Hosts = []string{cqlAddr}
		cfg.HostFilter = gocql.WhiteListHostFilter(cqlAddr)
		cfg.DisableInitialHostLookup = true
		return nil
	}
}
```

And:
```go
// GetSession returns CQL session to provided cluster.
func (s *Service) GetSession(ctx context.Context, clusterID uuid.UUID, opts ...SessionConfigOption) (session gocqlx.Session, err error) {
	s.logger.Info(ctx, "Get session", "cluster_id", clusterID)

	client, err := s.CreateClientNoCache(ctx, clusterID)
	if err != nil {
		return session, errors.Wrap(err, "get client")
	}
	defer logutil.LogOnError(ctx, s.logger, client.Close, "Couldn't close scylla client")

	cfg := gocql.NewCluster()
	for _, opt := range opts {
		if err := opt(ctx, clusterID, client, cfg); err != nil {
			return session, err
		}
	}
	// Fill hosts if they weren't specified by the options
	if len(cfg.Hosts) == 0 {
		sessionHosts, err := GetRPCAddresses(ctx, client, client.Config().Hosts)
		if err != nil {
			s.logger.Info(ctx, "GetSession", "err", err)
			if errors.Is(err, ErrNoRPCAddressesFound) {
				return session, err
			}
		}
		cfg.Hosts = sessionHosts
	}

	ni, err := client.AnyNodeInfo(ctx)
	if err != nil {
		return session, errors.Wrap(err, "fetch node info")
	}
	if err := s.extendClusterConfigWithAuthentication(clusterID, ni, cfg); err != nil {
		return session, err
	}
	if err := s.extendClusterConfigWithTLS(ctx, clusterID, ni, cfg); err != nil {
		return session, err
	}

	return gocqlx.WrapSession(cfg.CreateSession())
}
```

The annoying part here is how to avoid querying node info multiple times (or creating uncached client many times), but implementing #3892 should simplify that - with it we could even rewrite filling the TLS/Auth cluster config options with the same approach.

@karol-kokoszka What do you think about this idea (both in the future context of #3892, and the context of this PR)?

Reply:
This comment is not a blocker for the PR.
So, if you say that #3892 will simplify the logic, I agree.

Please check the one related to RAFT being supported by scylla server.